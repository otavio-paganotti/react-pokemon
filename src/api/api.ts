import axios, { AxiosInstance } from 'axios';

export const http: AxiosInstance = axios.create({
  baseURL: String(import.meta.env.VITE_APP_API_URL || ''),
});
