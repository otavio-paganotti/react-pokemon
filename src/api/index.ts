/**
 * Exporting the Axios Instance
 */
export * from './api';

/**
 * Exporting all modules
 */
export * as Pokemon from './services/rest/pokemon';
