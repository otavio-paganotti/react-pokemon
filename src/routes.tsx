import type { RouteObject } from 'react-router';
import MainLayout from '@/layouts/Default';
import Home from '@/pages/index/Index'
import Pokemon from '@/pages/pokemon/Index'

const routes: RouteObject[] = [
  {
    path: '*',
    element: <MainLayout />,
    children: [
      {
        index: true,
        element: <Home />,
      },
      {
        index: false,
        path: ':id',
        element: <Pokemon />
      }
    ],
  },
];

export default routes;