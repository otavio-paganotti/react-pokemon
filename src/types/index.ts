export type Nullable<T = unknown> = T | null;

export interface PokemonItem {
  name: string;
  url: string;
  id: number;
  content: Pokemon;
}

export interface Response<T> {
  count: number;
  next: Nullable<string>;
  previous: Nullable<string>;
  results: T[];
  pagination: {
    firstPage: number;
    lastPage: number;
    nextPage: number;
    prevPage: number;
    nextPages: number[];
    prevPages: number[];
    page: number;
  };
}

export interface Sprites {
  back_default: string;
}

export interface Stat {
  base_stat: number;
  effort: number;
  stat: Species;
}

export interface Type {
  slot: number;
  type: Species;
}

export interface VersionGroupDetail {
  level_learned_at: number;
  move_learn_method: Species;
  version_group: Species;
}

export interface Move {
  move: Species;
  version_group_details: VersionGroupDetail[];
}

export interface VersionDetail {
  rarity: number;
  version: Species;
}

export interface Species {
  name: string;
  url: string;
}

export interface HeldItem {
  item: Species;
  version_details: VersionDetail[];
}

export interface GameIndex {
  game_index: number;
}

export interface Ability {
  ability: Species;
  is_hidden: boolean;
  slot: number;
}

export interface Pokemon {
  abilities: Ability[];
  base_experience: number;
  forms: Species[];
  game_indices: GameIndex[];
  height: number;
  held_items: HeldItem[];
  id: number;
  is_default: boolean;
  location_area_encounters: string;
  moves: Move[];
  name: string;
  order: number;
  past_types: unknown[];
  species: Species;
  sprites: Sprites;
  stats: Stat[];
  types: Type[];
  weight: number;
}

export interface IPagination {
  firstPage: number;
  lastPage: number;
  nextPages: number[];
  prevPages: number[];
  nextPage: number;
  prevPage: number;
  page: number;
}
