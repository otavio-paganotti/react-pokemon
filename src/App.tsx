import routes from '@/routes';
import { useRoutes } from 'react-router-dom';
import '@/assets/css/main.scss';

const App = () => {
  const content = useRoutes(routes);

  return content;
};

export default App;
