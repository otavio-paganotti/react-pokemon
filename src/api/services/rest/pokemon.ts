import { http } from "@/api";
import { Pokemon, PokemonItem, Response } from "@/types";

export const getPokemon = async (id: number): Promise<Pokemon> => {
  const { data } = await http.get<Pokemon>(`/pokemon/${id}`);
  return data;
};

export const getPokemons = async (): Promise<Response<PokemonItem>> => {
  const { data } = await http.get<Response<PokemonItem>>("/pokemon");
  return data;
}