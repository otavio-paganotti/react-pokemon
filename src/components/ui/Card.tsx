import { PokemonItem } from "@/types";
import clsx from "clsx";
import { ReactNode } from "react";
import { Link } from 'react-router-dom';

interface IProps<T> {
  children?: ReactNode;
  item: T;
  noLink?: boolean;
}

const Card = <T extends PokemonItem>({ item, children, noLink }: IProps<T>) => {
  const getHP = () => item.content.stats?.find(stat => stat.stat.name === 'hp')?.base_stat;
  const getAttack = () => item.content.stats?.find(stat => stat.stat.name === 'attack')?.base_stat;
  const getDefense = () => item.content.stats?.find(stat => stat.stat.name === 'defense')?.base_stat;

  return (
    <div className="bg-white">
      <div className="flex justify-center">
        <img src={`https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/${item.id}.png`} />
      </div>

      <div className="p-4 flex flex-col gap-4">
        <div className="bg-primary">
          <div className="p-2 flex justify-between">
            <h2 className="font-bold capitalize">{item.name}</h2>
            <span>HP: {getHP()}</span>
          </div>
          <div className={clsx("px-2 pt-2", !noLink ? "pb-2" : "")}>
            <p>Ataque: {getAttack()}</p>
            <p>Defesa: {getDefense()}</p>
          </div>
          {children}
        </div>
        {!noLink && (<div className="flex justify-center">
          <Link to={`/${item.id}`} className="bg-primary text-white px-4">Access Pokemon</Link>
        </div>)}
      </div>
    </div>
  )
}

export default Card
