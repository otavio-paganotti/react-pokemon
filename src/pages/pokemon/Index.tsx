import { Pokemon as PokemonApi } from '@/api';
import Card from '@/components/ui/Card';
import { useLoader } from '@/hooks';
import { Pokemon } from '@/types';
import { useEffect } from 'react';
import { useParams } from 'react-router-dom';

export default function PokemonPage() {
  const params = useParams();

  const [loader, data, isLoading] = useLoader<Pokemon>({} as Pokemon);

  useEffect(() => {
    loader(() => PokemonApi.getPokemon(Number(params.id)));

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [params.id]);

  const getXAttack = () =>
    data.stats?.find((stat) => stat.stat.name === 'special-attack')?.base_stat;
  const getXDefense = () =>
    data.stats?.find((stat) => stat.stat.name === 'special-defense')?.base_stat;
  const getSpeed = () =>
    data.stats?.find((stat) => stat.stat.name === 'speed')?.base_stat;

  const getUrl = (id: number) =>
    `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${id}.png`;

  return (
    <>
      <div className="p-4 max-w-3xl mx-auto grid grid-cols-1 gap-4">
        <h1 className="text-4xl text-center">Pokemon</h1>
        <p className="text-center">This is the Pokemon page</p>
        {!isLoading && (
          <>
            <Card
              item={{
                id: data.id,
                url: getUrl(data.id),
                name: data.name,
                content: data,
              }}
              noLink={true}
            >
              <div className="px-2 pb-2">
                <p>Ataque especial: {getXAttack()}</p>
                <p>Defesa especial: {getXDefense()}</p>
                <p>Velocidade: {getSpeed()}</p>
              </div>
            </Card>
          </>
        )}
      </div>
    </>
  );
}
