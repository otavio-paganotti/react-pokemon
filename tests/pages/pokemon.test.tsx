import React from 'react';
import renderer, { act } from 'react-test-renderer';
import PokemonComponent from '../../src/pages/pokemon/Index';
import Card from '../../src/components/ui/Card';
import { describe, beforeEach, expect, test, vi } from 'vitest';
import { pokemon } from '../mocks/pokemon';
import { Pokemon } from '../../src/api';

vi.mock('../../src/api');
vi.mock('../../src/components/ui/Card');

const toJson = (component: renderer.ReactTestRenderer) => {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as renderer.ReactTestRendererJSON
};

vi.mock('react-router-dom', () => ({
  useParams: () => ({
    id: pokemon.id,
  }),
}));

const onMounted = async () => {
  vi.mocked(Pokemon.getPokemon).mockResolvedValueOnce(pokemon);

  let component;

  await act(() => {
    component = renderer.create(
      <PokemonComponent />
    )
  });

  return toJson(component);
};

describe('Pokemon.tsx', () => {
  let tree;

  beforeEach(async () => {
    vi.clearAllMocks();

    tree = await onMounted();
  });

  test('should render correctly the Pokemon component and take a snapshot', () => {
    expect(tree).toMatchSnapshot();
  });

  test('should have 1 h1 element', async () => {
    expect(tree.children).toHaveLength(2);

    expect(tree.children[0].type).toBe('h1');

    expect(tree.children[0].children).toHaveLength(1);

    expect(tree.children[0].children[0]).toBe('Pokemon');

    expect(tree.children[0].props).toEqual({
      className: 'text-4xl text-center',
    });
  });

  test('should have 1 Card child element must be called', async () => {    
    expect(Card).toHaveBeenCalled();
  });
});
