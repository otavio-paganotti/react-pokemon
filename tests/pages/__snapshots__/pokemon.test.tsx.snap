// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pokemon.tsx > should render correctly the Pokemon component and take a snapshot 1`] = `
<div
  className="p-4 max-w-3xl mx-auto grid grid-cols-1 gap-4"
>
  <h1
    className="text-4xl text-center"
  >
    Pokemon
  </h1>
  <p
    className="text-center"
  >
    This is the Pokemon page
  </p>
</div>
`;
